returnn/tf/layers/rec.py

```diff
@@ -2904,6 +2904,13 @@ def output_can_move_out(layer, _collocated=None):
         if layer in other_layer.dependencies:
           return False
         if other_layer.name in layer.collocate_with:
+          # layer is collocated with other_layer.
+          # we can move out layer if other_layer has no dependencies on any layer in loop but on the sibling layers
+          # that themselves are collocated with other_layer
+          layers_collocated_with_other_layer = [
+            sibling for sibling in layers_in_loop if other_layer.name in sibling.collocate_with]
+          assert layer in layers_collocated_with_other_layer
+          _collocated.update(layers_collocated_with_other_layer)
```

Comment:
See my added comment on the earlier code. I think calling `_collocated.update` directly without checking `output_can_move_out` is wrong.

We should have some similar code like `if layer.name in other_layer.collocate_with:` (the other way around). I think this would cover what you have here.

Also, I think we might/should move that `if layer in other_layer.dependencies:` below. And the `if other_layer in _collocated:` also after the `collocate_with` logic.

So I think this order:
```
if other_layer.name in layer.collocate_with: ...
if layer.name in other_layer.collocate_with: ...
if other_layer in _collocated: ...
if layer in other_layer.dependencies: ...
```